[WIP][benchmark] Flatten - Extended test family

palimondo opened the pull request:
The `Flatten` test family is based on recently added benchmarks `FlattenListLoop` and `FlattenListFlatMap` by @gottesmm. The original versions had unnecessarily large base workloads, which prevented more precise measurement, so their base workload was lowered by a factor of 20. See discussion on https://github.com/apple/swift/pull/20116#issuecomment-434894826

Since these are recent additions to the Swift Benchmark Suite, I’m removing the originals and reintroducing them under new names `Flatten.Array.Tuple4.flatMap` and `Flatten.Array.Tuple4.for-in.Reserve`without going through the [`legacyFactor`](https://github.com/apple/swift/pull/20212), along with extended performance test coverage inspired by these two benchmarks.

The implementation of these benchmarks has evolved substantially with @atrick's help ([see conversation](https://github.com/apple/swift/pull/20552#discussion_r236265215)). Thank you Andrew!

---

The `Flatten` benchmark family tests the performance of `flatMap` function and functionally equivalent `map.joined`, along with their lazy variants relative to an imperative approach with simple for-in loop across a selection of representative types.

For transforming fully materialized collection with contiguous memory layout additional Unsafe versions were created as attempts at manual optimization that try to eliminate the abstraction overhead using assumptions about the internal memory layout of underlying data structures.

## Colors

First hypotetical scenario is transforming an array of RBGA pixel values represented as `struct ColorVal { let r, g, b, a: UInt8 }` to flat `[UInt8]` in ARGB format. In case of `[ColorVal]`, this means that the real work being performed is copying of byte swizzled raw memory, obfuscated by type casting and higher-level abstractions (structs, arrays).

The alternative type `class ColorRef` demonstrates the impact of using reference type.

After experimenting with Unsafe variants, conforming the `ColorVal` to `Sequence` protocol by creating a custom iterator, which performes the color component swizzling, showed promising performance (better than imperative approach). That varaint is called `SwizSeq`. Turns out that conforming the type to `Collection` protocol (in `SwizCol` variant) is even better, allowing compiler to optimize the lazy variants for almost 4x gain, beating even the best performing Unsafe variant that copies the colors byte-by-byte.

See http://wiki.c2.com/?SufficientlySmartCompiler

The Unsafe variants were originally meant as aspirational goals for the functional approach, but are now kept here as artifacts of partial improvements over the imperative approach, which demonstrate unexpected performance behavior.

## Tuple4 and Array4

Second scenario tests the performance of flattening the compound type `(Int, Int, Int, Int)`, typealiased as `Tuple4`, into `[Int]`. This variant compensates for the larger data type by ommiting the structural transformation. In case of fully materialized collection, `[Tuple4]`, the real work is simply a type cast. There's currently no Array API to perform this in O(1), pending [SE-0223](https://forums.swift.org/t/15194/41). Therefore the Unsafe variants perform a simple memory copy.

Next variant, `Array4` uses 4 element "static" array instead of the `Tuple4`, and is meant do demonstrate the relative cost of switching to this currency type. This is important, because Array is naturally used, thanks to its syntactic sugar, as the flattened collection in all the functional-style scenarios, i.e. the closures in `flatMap` and `map.joined` are producing "static" arrays on the fly.

The `Tuple4` and `Array4` type groups are varied across 3 container types:

* `Flatten.Array` is a fully materialized collection, 
* `Flatten.LazySeq` is lazily generated sequence, and 
* `Flatten.AnySeq.LazySeq` is a type erased version of the latter.

After [SE-0234](https://forums.swift.org/t/18002), no standard library API returns `AnySequence` anymore, but the tests `Flatten.LazySeq.Tuple4.flatMap` and `Flatten.AnySeq.LazySeq.Tuple4.flatMap` hint at the hidden potential for improvement in the utter performance debacle that is `Flatten.LazySeq`. The `AnySeq` group could be removed in the future, when _that_ deoptimization is properly addressed.

---

The tests follow naming convention proposed in https://github.com/apple/swift/pull/20334

palimondo committed:
[benchmark] Adjust Driver’s console output format

…to handle longer benchmark names, assuming maximum length of 40 characters.

palimondo committed:
[benchmark] LogParser: Accept `-` in bench. names

Extend parser to support benchmark names that include `-` in names, as proposed in PR #20334.

palimondo committed:
[benchmark] Flatten - Extended test family

The extended `Flatten` test family is based on recently added benchmarks `FlattenLostLoop` and `FlattenListFlatMap`. They had unnecessarily large base workloads, which prevented more precise measurement. Their base workload was lowered by a factor of 20. See discussion on https://github.com/apple/swift/pull/20116#issuecomment-434894826

Since these are recent additions to the Swift Benchmark Suite, I’m removing the originals and reintroducing them under new names `Flatten.Array.Tuple4.flatMap` and `Flatten.Array.Tuple4.for-in.Reserve`without going through the `legacyFactor`.

Based on these two templates, this commit introduces thorough performance test coverage of the related space including:

* method chain `map.joined`
* naive for-in implementation without `reserveCapacity`
* few Unsafe variants that should serve as aspirational targets for ideally optimized code
* lazy variants
* variants for different underlying types: 4 element Array, struct and class in addition to the original 4 element tuple
* variants that flatten Sequence instead of Array

The tests follow naming convention proposed in https://github.com/apple/swift/pull/20334

palimondo commented:
@gottesmm @atrick @eeckstein @airspeedswift Please run benchmark and review.
I have added more variants than is necessary, for you to choose here. I'll remove the excess in a followup commit once we agree on what's valuable to keep.

I'm guessing the whole `AnySeq` group could be dropped, since it adds no actionable info and `AnySequence` will not be a concern once the [SE-0234](https://forums.swift.org/t/se-0234-removing-sequence-subsequence/17750/18) passes. Only surprising thing is `Flatten.Seq.Tuple4.flatMap` vs. `Flatten.AnySeq.Seq.Tuple4.flatMap` where the type erased version actually **improves from 8270 to 1855**, hinting at the hidden potential in the utter performance debacle that is `Flatten.Seq`. I'm guessing there is some serious de-optimization going on somewhere, because when I've profiled it, the fast one is dominated by	 	        `_platform_memmove$VARIANT$Base`, while the slow one drowns in `_swift_release_dealloc, swift_deallocClassInstance, objc_destructInstance, _object_remove_assocations`. The same affliction affects the `Flatten.Array.Tuple4.lazy.flatMap`, while non-lazy version flies with `memmove`.

I was struggling to come up with proper Unsafe implementation for the `Flatten.Array.Color.Val` that would be an equivalent to the aspirational speed-of-light goal `Flatten.Array.Tuple4.Unsafe` we've cooked-up with @atrick in https://github.com/apple/swift/pull/20116#issuecomment-437258412. I know these are all "wrong", because they are x86 specific. I need help there.

I was also surprised by the slowdown between `Tuple4` and `Array4`… maybe Swift could use some compile-time static array optimizations?

Additionally I was thinking about adding an `ArrayN8` variant, that would include the same number of flattened elements as `Array4`, but the content of the outer array would be composed of arrays with variable lengths: 2, 2, 4, 8 (maybe up to 16) - repeating… would that make sense?

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+    let input = arrayColorVals
+
+    let f = input.withUnsafeBufferPointer {
+      $0.baseAddress!.withMemoryRebound(to: UInt32.self, capacity: input.count) {
```
Why are you viewing this at `UInt32` and doing bitwise arithmetic instead of viewing it as `UInt8`?

palimondo replied:
As is described in the `.gyb` file, the pretend scenario for `Color` group is conversion from RGBA to ARBG pixel format. To do the rotation, I have to step in 32bit chunks. Or is there other way?

atrick replied:
I get that an RGBA->ARGB conversion is happening, although I would never guess from the function name and I don't see any comments.

I would have just used `withUnsafeBytes` instead of `withMemoryRebound` and looped over the indices `for i on stride(from: 0, to: N, by: 4)`.

atrick replied:
If there's some reason that you decided to write it one way versus another, maybe because you're trying to emulate the non-unsafe loop, then please comment that!

palimondo replied:
The idea was to systematically extend the coverage beyond the `Int` tuple @gottesmm used. So I’ve added array of arrays, then I though I need to also cover struct and a class. To make it somewhat more realistically motivated, I thought of swizzling the byte order of pixel data. That also seems like a more realistically motivated scenario for the aspirational Unsafe version than the contrived `Tuple4.Unsafe` variant.

Can you please write the swizzling using `stride` for me? I don’t understand how to apply what you suggested.

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+          (rawColor: UInt32) -> [UInt8] in
+          let rgba = rawColor.bigEndian
+          let argb = (rgba << 24) | (rgba >> 8) // rotate
+          return stride(from:24, through:0, by: -8).map {
```
This would be so much easier to read of the shift operations are just written out explicitly.

palimondo replied:
I don’t know what you mean. This is as explicit as I could do it. I’m asking for help with the `Color.Unsafe`, because I don’t like any of the 3 I did. None feels canonical.

atrick replied:
I mean:
 `[ UInt8(truncatingIfNeeded: argb >> 24,
     UInt8(truncatingIfNeeded: argb >> 16,
 ...]`

Why are there two versions of `run_FlattenArrayColorValUnsafe2`?

If you insist on loading `UInt32` then loading `rgba.bigEndian` should rectify that in a platform independent way right?

palimondo replied:
I don’t insist on anything here. This version performs orders of magnitude slower than naive for-in loop. It’s here as a witness of how I’ve failed to apply the pattern you’ve taught me with `Unsafe.Tuple4` to this new byte swizzling scenario. As I said, this is my first Unsafe rodeo…

palimondo replied:
That’s basically what `Flatten.Array.Color.Val.flatMap` does. I wouldn’t expect this to gain any performance advantage from going through Unsafe. So no. My goal with Unsafe variant is to demonstrate the fastest possible way to get to the `[UInt8]` by skipping as much overhead as possible by leveraging knowledge about `ColorVal`’s memory layout etc. Unless it beats imperative for-in by a good margin, there’s no point in including it. It should establish a goal for best-case optimized `flatMap`.

atrick replied:
This is why I asked you to explain why you're adding each test case. I have no idea what overhead you think you're removing in each case.

Do you have a faster implementation than the one I posted?

palimondo replied:
> If you insist on loading UInt32 then loading rgba.bigEndian should rectify that in a platform independent way right?

I'm not sure I'm using the right assumptions here. As I said, I had to use the `.bigEndian` conversion here, to make it work correctly (by trial and error). But I doubt this achieves true platform independence… You said in https://github.com/apple/swift/pull/20116#issuecomment-437258412 regarding the `Tuple4.Unsafe`:
> Note: Swift does not provide any formal layout rule that says an array of homogeneous tuples is laid out the same as an array of those tuple elements. But with builtin integer types I feel that you're safe making this assumption.

So I think the platform dependent and totally un-guaranteed assumption I'm making here is about the struct endianness or the order of the 4 `Int`s packed in the `struct ColorVal`. Which makes me think that this particular benchmark should be enclosed in `#if #else` block to be x86 specific for now… or is there some layout guarantee I can base this on?

> Do you have a faster implementation than the one I posted?

Unsafe2 and 3 are both faster than yours (Unsafe4-atrick), which is only slightly faster than `flatMap` and worse than the naive imperative approach.

### Flatten.Array.Color.Val
| Variant | Min | Q1 | Med | Q3 | Max |
|---|---:|---:|---:|---:|---:|
| Unsafe | 9622 | 9843 | 10056 | 10396 | 13497 |
| Unsafe2 | 290 | 290 | 291 | 291 | 296 |
| Unsafe3 | 352 | 354 | 355 | 365 | 445 |
| Unsafe4-atrick | 517 | 518 | 526 | 534 | 603 |
| flatMap | 551 | 552 | 559 | 565 | 580 |
| for-in.Naive | 465 | 465 | 477 | 893 | 1016 |
| lazy.flatMap | 7098 | 7180 | 7273 | 7440 | 7704 |
| lazy.map.joined | 7073 | 7207 | 7334 | 7507 | 8096 |
| map.joined | 9768 | 9956 | 10237 | 10511 | 11760 |

palimondo replied:
I have thought about it some more and came up with yet another conversion strategy: creating custom `ColorVal` conformance to `Sequence` protocol by implementing `SwizzlingIterator`.

<details>

````swift
extension ColorVal {
  public struct SwizzlingIterator {
    let c: ColorVal
    var i: UInt8 = 0
    init(_ c: ColorVal) { self.c = c }
  }
}

extension ColorVal.SwizzlingIterator: IteratorProtocol {
  public mutating func next() -> UInt8? {
    i = i &+ 1
    switch i {
    case 1: return c.a
    case 2: return c.r
    case 3: return c.g
    case 4: return c.b
    default: return nil
    }
  }
}

extension ColorVal: Sequence {
  public func makeIterator() -> SwizzlingIterator {
    return SwizzlingIterator(self)
  }
}
````
</details>

These are the performance results for following variants:

| Variant | Min | Q1 | Med | Q3 | Max | Code |
|---|---:|---:|---:|---:|---:|---|
| flatMap.SwizzleIt | 564 | 573 | 580 | 602 | 692 | `input.flatMap { $0 }` |
| lazy.flatMap.SwizzleIt | 319 | 319 | 319 | 319 | 319 | `input.lazy.flatMap { $0 }` |
| joined.SwizzleIt | 319 | 319 | 319 | 319 | 319 | `input.joined()` |
| lazy.joined.SwizzleIt | 319 | 319 | 319 | 319 | 319 | `input.lazy.joined()` |

What do you make of this? Is this worth including? It looks like the last three did produce the exact same code after optimization...

I have also tried `flatMap` and `map.joined` using `ContiguousArray([$0.a, $0.r, $0.g, $0.b])` in the closure, expecting to see some improvement for `Color.Ref`, but saw no significant change compared to plain `Array`. Was I using it right with the array literal syntax?

palimondo replied:
That's it. Byte-By-Byte: **136**! Thank you Andrew.

palimondo replied:
>  If you find something faster then this then I'd be interested to know why

😏 Building up from `SwizzleIt` from above... conforming ColorVal to `RandomAccessCollection` gives runtime of **82** for `joined`, `lazy.flatMap` and `lazy.joined`

````swift
extension ColorVal: RandomAccessCollection {
  public var count: Int { return 4 }
  public var startIndex: Int { return 0 }
  public var endIndex: Int { return 4 }
  public func index(after i: Int) -> Int { return i &+ 1 }
  public subscript(position: Int) -> UInt8 {
    switch position {
    case 0: return a
    case 1: return r
    case 2: return g
    case 3: return b
    default: return 0 // never happens
    }
  }
}
````
Profiling this in instruments, the assembler code looks ordinary. So maybe it could be even faster, if it did swizzling using vector instructions?

palimondo replied:
Hmm... 🧐🤨 I've been exploring the disassembled code of the fastest variant over at https://swift.godbolt.org/ … and now I'm not sure how much sense it makes, the way I've been comparing lazy versions with eager ones. When I've switched to unsafe addition for counting the elements in the `CountLast`, that record time 82 went down to 51. But of course it's fast to get to the last element by exhausting the sequence, when you don't have to do anything else... But on the other hand, switching to fully materializing the lazy sequences with `Array.init(Sequence)` gives unfair advantage to the eager versions and muddles the water by mixing in the Array's sequence/collection initializer with it's optimization baggage. I'll explore dumping the sequence into preallocated `UnsafeBufferPointer`.

palimondo replied:
It looks like the improvement potential from using inlined lazy sequences and collections is real. I've replaced the `CountLast` with `materializeSequence`:

````swift
struct CountLast<T> {
  var count: Int = 0
  var last: T? = nil
  static func collect(_ cl: inout CountLast, _ element: T){
    cl.count += 1
    cl.last = element
  }
}

// Helper function to fully materialize Sequence into preallocated buffer.
// Cheats by knowing `flatSize` in advance. But demonstrates the power of
// inlining the Sequence initializer.
func materializeSequence<S: Sequence>(_ s: S) -> [S.Element]
  where S.Element: Initializable {
  var a = ContiguousArray<S.Element>(repeating: S.Element(), count: flatSize)
  a.withUnsafeMutableBufferPointer { buffer in
    var i = 0
    _ = s.reduce(into:buffer) {
      $0[i] = $1
      i &+= 1
    }
  }
  return Array(a)
}
````


<details open>
  <summary><strong>Improvement (6)</strong></summary>
  
TEST                                     | OLD   | NEW   | DELTA  | RATIO     
---                                      | ---   | ---   | ---    | ---       
Flatten.ColorVal.lazy.map.joined.SwizSeq | 319   | 234   | -26.6% | **1.36x** 
Flatten.ColorVal.lazy.flatMap.SwizSeq    | 319   | 234   | -26.6% | **1.36x** 
Flatten.ColorVal.lazy.flatMap.SwizCol    | 82    | 63    | -23.2% | **1.30x** 
Flatten.ColorVal.lazy.map.joined.SwizCol | 82    | 63    | -23.2% | **1.30x** 
Flatten.ColorVal.map.joined.SwizSeq      | 540   | 488   | -9.6%  | **1.11x** 
Flatten.ColorVal.map.joined.SwizCol      | 291   | 277   | -4.8%  | **1.05x (?)** 

</details>

<details>
  <summary>Regression (8)</summary>
  
TEST                                     | OLD   | NEW   | DELTA  | RATIO     
---                                      | ---   | ---   | ---    | ---       
Flatten.AnySeq.LazySeq.Tuple4.flatMap    | 1001  | 1172  | +17.1% | **0.85x** 
Flatten.LazySeq.Tuple4.for-in            | 942   | 1062  | +12.7% | **0.89x** 
Flatten.AnySeq.LazySeq.Tuple4.for-in     | 975   | 1083  | +11.1% | **0.90x** 
Flatten.LazySeq.Array4.for-in            | 8103  | 8725  | +7.7%  | **0.93x (?)** 
Flatten.ColorVal.map.joined.Array        | 10575 | 11258 | +6.5%  | **0.94x (?)** 
Flatten.Array.Array4.lazy.flatMap        | 1297  | 1377  | +6.2%  | **0.94x** 
Flatten.Array.Array4.lazy.joined         | 1298  | 1376  | +6.0%  | **0.94x** 

</details>
<details >
  <summary>No Changes (26)</summary>
  
TEST                                     | OLD   | NEW   | DELTA  | RATIO     
---                                      | ---   | ---   | ---    | ---       
Flatten.AnySeq.LazySeq.Tuple4.map.joined | 11280 | 11469 | +1.7%  | 0.98x (?) 
Flatten.Array.Array4.joined              | 309   | 309   | +0.0%  | 1.00x     
Flatten.Array.Tuple4.lazy.flatMap        | 7058  | 7203  | +2.1%  | 0.98x (?) 
Flatten.Array.Tuple4.lazy.map.joined     | 7048  | 7199  | +2.1%  | 0.98x (?) 
Flatten.Array.Tuple4.map.joined          | 10094 | 10090 | -0.0%  | 1.00x (?) 
Flatten.ColorRef.lazy.flatMap.Array      | 8040  | 8042  | +0.0%  | 1.00x (?) 
Flatten.ColorRef.lazy.flatMap.ContArr    | 7908  | 7998  | +1.1%  | 0.99x (?) 
Flatten.ColorRef.lazy.flatMap.SwizCol    | 1072  | 1123  | +4.8%  | 0.95x     
Flatten.ColorRef.lazy.flatMap.SwizSeq    | 1242  | 1278  | +2.9%  | 0.97x     
Flatten.ColorRef.lazy.map.joined.Array   | 7996  | 8072  | +1.0%  | 0.99x (?) 
Flatten.ColorRef.lazy.map.joined.ContArr | 7852  | 8064  | +2.7%  | 0.97x (?) 
Flatten.ColorRef.lazy.map.joined.SwizCol | 1072  | 1117  | +4.2%  | 0.96x     
Flatten.ColorRef.lazy.map.joined.SwizSeq | 1246  | 1281  | +2.8%  | 0.97x     
Flatten.ColorRef.map.joined.Array        | 11859 | 11829 | -0.3%  | 1.00x (?) 
Flatten.ColorRef.map.joined.ContArr      | 11437 | 11525 | +0.8%  | 0.99x (?) 
Flatten.ColorRef.map.joined.SwizCol      | 2347  | 2395  | +2.0%  | 0.98x (?) 
Flatten.ColorRef.map.joined.SwizSeq      | 2538  | 2512  | -1.0%  | 1.01x (?) 
Flatten.ColorVal.lazy.flatMap.Array      | 7056  | 7074  | +0.3%  | 1.00x (?) 
Flatten.ColorVal.lazy.flatMap.ContArr    | 7092  | 6981  | -1.6%  | 1.02x (?) 
Flatten.ColorVal.lazy.map.joined.Array   | 7094  | 7035  | -0.8%  | 1.01x (?) 
Flatten.ColorVal.lazy.map.joined.ContArr | 7050  | 7034  | -0.2%  | 1.00x (?) 
Flatten.ColorVal.map.joined.ContArr      | 10313 | 10391 | +0.8%  | 0.99x (?) 
Flatten.LazySeq.Array4.flatMap           | 7304  | 7601  | +4.1%  | 0.96x (?) 
Flatten.LazySeq.Array4.joined            | 7308  | 7570  | +3.6%  | 0.97x (?) 
Flatten.LazySeq.Tuple4.flatMap           | 7298  | 7417  | +1.6%  | 0.98x (?) 
Flatten.LazySeq.Tuple4.map.joined        | 7298  | 7516  | +3.0%  | 0.97x (?) 

</details>

These regressions don't concern me much, because it looks like using `Array` in the closures is generally some kind of an optimization barrier...

I have initially experimented with materializing the sequences using Array's initializer (`let fa = Array(f)`), but it was never faster than the imperative approach. Which surprised me, because`.lazy` is the performance magic, [in my experience](https://forums.swift.org/t/troubling-sequence/13130)... The old comment next to template using `CountLast` justified its necessity for good performance, by speculating that the measured difference was due to memory allocation overhead. But `materializeSequence` disproves this theory... So this is a bit confusing now. 

Array's sequence initializer seems to work well for `Tuple4.Unsafe.InitSeq` and also for `ColorVal` variants `Unsafe.UInt32InitSeq` and `Unsafe.ColorValInitSeq`. But the likes of `ColorVal.lazy.flatMap.SwizCol` are not getting the boost I'd expect when plugged into `Array.init<Sequence>`. They are clearly best with `CountLast` and also with `materializeSequence`, even better than `Unsafe.BytesReserve` (139 μs)!

It looks like some performance bug to me… @atrick how would you handle this? Do we need to fix this before ABI freeze? 
// cc @airspeedswift

#### `materializeSequence(f)` vs. `Array(f)`
<details>
  <summary>Regression (17)</summary>
  
TEST                                     | OLD   | NEW   | DELTA    | RATIO     
---                                      | ---   | ---   | ---      | ---       
Flatten.ColorVal.lazy.map.joined.SwizCol | 63    | 716   | +1036.5% | **0.09x** 
Flatten.ColorVal.lazy.flatMap.SwizCol    | 63    | 675   | +971.4%  | **0.09x** 
Flatten.ColorVal.lazy.map.joined.SwizSeq | 234   | 891   | +280.8%  | **0.26x** 
Flatten.ColorVal.lazy.flatMap.SwizSeq    | 234   | 850   | +263.2%  | **0.28x** 
Flatten.ColorVal.map.joined.SwizCol      | 277   | 677   | +144.4%  | **0.41x** 
Flatten.ColorVal.map.joined.SwizSeq      | 488   | 899   | +84.2%   | **0.54x** 
Flatten.LazySeq.Array4.flatMap           | 7601  | 9100  | +19.7%   | **0.84x (?)** 
Flatten.ColorRef.lazy.map.joined.SwizCol | 1117  | 1326  | +18.7%   | **0.84x** 
Flatten.ColorRef.lazy.flatMap.SwizCol    | 1123  | 1325  | +18.0%   | **0.85x** 
Flatten.Array.Array4.lazy.flatMap        | 1377  | 1624  | +17.9%   | **0.85x** 
Flatten.Array.Array4.lazy.joined         | 1376  | 1608  | +16.9%   | **0.86x** 
Flatten.ColorRef.map.joined.SwizSeq      | 2512  | 2778  | +10.6%   | **0.90x** 
Flatten.LazySeq.Tuple4.flatMap           | 7417  | 8189  | +10.4%   | **0.91x (?)** 
Flatten.LazySeq.Tuple4.map.joined        | 7516  | 8252  | +9.8%    | **0.91x** 
Flatten.ColorRef.lazy.flatMap.SwizSeq    | 1278  | 1384  | +8.3%    | **0.92x** 
Flatten.ColorRef.lazy.map.joined.SwizSeq | 1281  | 1383  | +8.0%    | **0.93x** 
Flatten.ColorRef.map.joined.SwizCol      | 2395  | 2551  | +6.5%    | **0.94x** 

</details>

<details >
  <summary>Improvement (4)</summary>
  
TEST                                     | OLD   | NEW   | DELTA    | RATIO     
---                                      | ---   | ---   | ---      | ---       
Flatten.AnySeq.LazySeq.Tuple4.flatMap    | 1172  | 1005  | -14.2%   | **1.17x** 
Flatten.LazySeq.Tuple4.for-in            | 1062  | 938   | -11.7%   | **1.13x** 
Flatten.AnySeq.LazySeq.Tuple4.for-in     | 1083  | 969   | -10.5%   | **1.12x** 
Flatten.LazySeq.Array4.for-in            | 8725  | 8234  | -5.6%    | **1.06x (?)** 

</details>

<details >
  <summary>No Changes (18)</summary>
  
TEST                                     | OLD   | NEW   | DELTA    | RATIO     
---                                      | ---   | ---   | ---      | ---       
Flatten.AnySeq.LazySeq.Tuple4.map.joined | 11469 | 11691 | +1.9%    | 0.98x (?) 
Flatten.Array.Array4.joined              | 309   | 309   | +0.0%    | 1.00x     
Flatten.Array.Tuple4.lazy.flatMap        | 7203  | 7455  | +3.5%    | 0.97x (?) 
Flatten.Array.Tuple4.lazy.map.joined     | 7199  | 7511  | +4.3%    | 0.96x (?) 
Flatten.Array.Tuple4.map.joined          | 10090 | 10108 | +0.2%    | 1.00x (?) 
Flatten.ColorRef.lazy.flatMap.Array      | 8042  | 8208  | +2.1%    | 0.98x (?) 
Flatten.ColorRef.lazy.flatMap.ContArr    | 7998  | 8175  | +2.2%    | 0.98x (?) 
Flatten.ColorRef.lazy.map.joined.Array   | 8072  | 8218  | +1.8%    | 0.98x (?) 
Flatten.ColorRef.lazy.map.joined.ContArr | 8064  | 8216  | +1.9%    | 0.98x (?) 
Flatten.ColorRef.map.joined.Array        | 11829 | 12089 | +2.2%    | 0.98x (?) 
Flatten.ColorRef.map.joined.ContArr      | 11525 | 11837 | +2.7%    | 0.97x (?) 
Flatten.ColorVal.lazy.flatMap.Array      | 7074  | 7247  | +2.4%    | 0.98x (?) 
Flatten.ColorVal.lazy.flatMap.ContArr    | 6981  | 7172  | +2.7%    | 0.97x (?) 
Flatten.ColorVal.lazy.map.joined.Array   | 7035  | 7226  | +2.7%    | 0.97x (?) 
Flatten.ColorVal.lazy.map.joined.ContArr | 7034  | 7225  | +2.7%    | 0.97x (?) 
Flatten.ColorVal.map.joined.Array        | 11258 | 10976 | -2.5%    | 1.03x (?) 
Flatten.ColorVal.map.joined.ContArr      | 10391 | 10747 | +3.4%    | 0.97x (?) 
Flatten.LazySeq.Array4.joined            | 7570  | 7664  | +1.2%    | 0.99x (?) 

</details>

atrick replied:
> It looks like some performance bug to me… @atrick how would you handle this? Do we need to fix this before ABI freeze?
// cc @airspeedswift

We can replace `@inlinable` stdlib implementations with semantically equivalent ones without breaking ABI. We can also theoretically make things `@inlinable` that weren't before. Without debugging it further, I think this is just a normal bug against the optimizer, or rather an opportunity for better performance.

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+          // rotate (inverted direction, because x86 is little endian)
+          (rgba >> 24) | (rgba << 8)
+        }
+        .withUnsafeBytes(Array.init)
```
[comment] As before, this code is totally obscure. Explain that it's constructing an Array of `UInt8` using the `Sequence` initializer.

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+    let input = arrayColorVals
+
+    let f = input.withUnsafeBufferPointer {
+      $0.baseAddress!.withMemoryRebound(to: UInt32.self, capacity: input.count) {
```
80-col violation.

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+    let input = arrayColorVals
+
+    let f = input.withUnsafeBufferPointer {
+      $0.baseAddress!.withMemoryRebound(to: UInt32.self, capacity: input.count) {
```
80-col violation.

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+    let f = input.withUnsafeBufferPointer {
+      $0.baseAddress!.withMemoryRebound(to: UInt32.self, capacity: input.count) {
+        UnsafeBufferPointer(start:$0, count: input.count).map { rgba in
+          // rotate (inverted direction, because x86 is little endian)
```
I'm not sure why you want x86 specific benchmark logic. You could view the memory as `UInt8` or just not include this nonprotable case.

atrick commented on benchmark/single-source/Flatten.swift:
```diff
@@ -0,0 +1,962 @@
+    let input = arrayColorVals
+
+    let f = input.map { c in ColorVal(r: c.a, g: c.r, b: c.g, a: c.b) }
+        .withUnsafeBytes(Array.init)
```
[comment] Brief comment explaining what this does and why it's interesting.

atrick reviewed (commented):
For future benchmark mainenance, it's important to understand the
justification and intention behind the benchmarks. Please add comments
to test cases explaining what is being tested and why we care about
performance of this particular test vs. more conventional ways to
write it. I know nobody has really done this in the past, but I think
it's important, especially when adding such a large number of
strangely written tests.

Aside from that, I reviewed the Unsafe benchmarks and they look correct, although they are hard to read without comments.

eeckstein commented on benchmark/single-source/Flatten.swift.gyb:
```diff
@@ -0,0 +1,275 @@
+//
+//===----------------------------------------------------------------------===//
+
+% # Ignore the following warning. This _is_ the correct file to edit.
```
Please don't add gyb files. Eventually we want to move the swift bm suite to swiftpm and then we cannot support gyb files anymore. Beside that, gyb files make it harder to understand and maintain the code.

palimondo replied:
I’m really surprised by this perspective. I find that to get proper performance test coverage which tests majority of legal combinations from stdlib types and methods, GYB is indispensable especially for the for ease of maintenance. The only alternative is copy & paste of some template and manual modification, which would be much more error prone.

Generally speaking, currently Swift’s biggest performance weakness is optimizer fragility. Small variations in expressing functionally  equivalent code often leed to unexpected pitfalls with orders of magnitude worse performance. My idea to harden the optimizer is to produce broad benchmark coverage and file bugs for all the gotchas. Is that a bad approach?

I’m currently working on reintroducing Existential benchmark family and the first step was to create a `.gyb` that let’s me do same refactoring without tearing my hair out in the 800 LOC file.

eeckstein replied:
There are 2 things here:
1) gyb: I'm strongly against adding more gyb files for the reasons I explained above. Eventually we will also eliminate the existing gyb files (and BTW, that's also what the stdlib team is doing).

2) Adding a lot of benchmarks just to cover all combinations is a problematic approach and not workable in the long term, because it will result in long benchmark runtimes. In general I would prefer that we add fewer but more "relevant" benchmarks (whatever this means). For example, simple operations, which we expect to be optimized to a trivial code pattern, are ofter better tested with a lit test than a benchmark. On the other hand, complex operations often don't need many combinations to be benchmarked, e.g. stdlib's sort can be benchmarked with a few types/array sizes - there is not much value in exploding the whole problem space.

But: what can be done is to add a large set of benchmarks for a specific feature, which are disabled by default (with tags). Whenever someone works on that feature he/she can run this large set locally to fully verify the performance.

palimondo replied:
> Eventually we want to move the swift bm suite to swiftpm and then we cannot support gyb files anymore.

Building benchmarks with swiftpm does not technically conflict with use of GYB, because we are not generating the swift files as a part of cmake build. These are generated manually by running `generate_harness.py`. That  is why we are also committing the generated `.swift` files to git, design mandated by @gottesmm when we added the support for GYB, specifically to accommodate the future swiftpm builds.

> Eventually we will also eliminate the existing gyb files (and BTW, that's also what the stdlib team is doing).

The stdlib team is doing that, because the improved expressivity of Swift has obviated the need to generate a lot of boilerplate. This argument does not apply to maintenance of benchmark families build from common template that need to be parametrized for different variants.

> Adding a lot of benchmarks just to cover all combinations is a problematic approach and not workable in the long term, because it will result in long benchmark runtimes.

This is simply not true. See #20666:
> New BenchmarkCategory.existential was added to tag these tests. Running these 108 test in a manner equivalent to run_smoke_bench (time ./Benchmark_O --tags=existential --sample-time=0.0025 --num-samples=3) take 1.3 seconds on my 2008 MBP — properly sized benchmarks improve measurement precision and have negligible impact on the overall time it takes to run benchmark suite.

palimondo replied:
@eeckstein:
> Adding a lot of benchmarks just to cover all combinations is a problematic approach and not workable in the long term, because it will result in long benchmark runtimes.

This whole new `Flatten` family at the current state (before trimming all the fat):
````
$ time ./Benchmark_O --num-samples=3 --sample-time=0.0025 {319..386}
…
Total performance tests executed: 68

real	0m1.298s
user	0m0.949s
sys	0m0.045s
````
On a 2008 MBP.

palimondo commented:
Could somebody please run benchmarks, so that we don’t talk in the abstract?

atrick commented:
@swift-ci benchmark.

palimondo commented:
Is it possible to re-run the benchmarks with #20667, or does that have to land first?

palimondo committed:
[benchmark] Flatten: fine-tuning and documentation

* Documented the motivation behind the various test scenarios.
* More descriptive names for Unsafe variants.
* New groups for color swizzling using Sequence and Collection conformances.
* Reduced number of Array4 and Tuple4 variants.

palimondo commented:
@atrick Please have a look at the latest version and re-run benchmarks with #20667.

palimondo committed:
[benchmark] Flatten: materializeSequence

For better comparability between eager and lazy results, fully materialize the lazy sequences into array using helper function.

The `materializeSequence` still beats the performance of Array.init<Sequence> for lazy sequences in ColorVal group by an order of magnitude.

palimondo commented:
Please benchmark with following pull request:
apple/swift#20667

@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@eeckstein Can you please run benchmark? 🤖 still ignores me:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please test

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please test

eeckstein commented:
@swift-ci benchmark

eeckstein commented:
@swift-ci benchmark

palimondo commented:
I understand that at a first glance the above benchmark report looks just like a lot of numbers, so let's try to tell some story around `-O` build performance with a proper tooling (_manual table prototype_) that builds on #20334. I'm very happy with how `Colors` turned out here. I'll return to `Tuple4` and `Array4` later.

I understand that at a first glance the above report looks just like a lot of numbers, so let's try to tell a story around this with a proper tooling (manual table prototype) that builds on #20334. I'm very happy with how `Colors` turned out here. I'll return to `Tuple4` and `Array4` later.

The `ColorVal` table is divided into 3 parts by the implementation of the `RGBA -> ARGB` transformation: functional at the top, `Unsafe` in the middle and imperative at the bottom (`—`). They are sorted in ascending order of the measured time in the first numeric column. The functional part is further divided into `flatMap` and `map.joined` with their relative comparison in-between (<sup>m.j/fM</sup>) and into 2 column sub-groups: `lazy` and eager (`—`) again separated by their relative comparison (<sub>—/l</sub>). When the ratio of compared results is in 0.7—1.3 range, it has no special format. Slowdowns above 1.3 are in<sub> _italics_</sub> and speedups below 0.7 are in <sub>**bold**</sub>.

Each sub-group is further split into 4 more columns for the variants: the type/protocol used to perform the transformation:
* `Collection` — **`SwizCol`**
* `Sequence` — **`SwizSeq`**
* **`Array`**
* `ContiguousArray` — **`ContArr`**
These are again sorted in ascending order by the fastest variant in the fastest sub-group.

<table>
<caption><h3 align="center"><code>ColorVal
<thead>
<tr><td rowspan="5"><th colspan="4"><code>lazy<th><th colspan="4"><code>—
<tr>
<th align="right"><code>SwizCol <th><th rowspan="2"><th rowspan="3">
<th rowspan="3">
<th align="right"><code>SwizCol <th><th rowspan="2"><th rowspan="3">
<tr>
<th colspan="2" align="right"><code>SwizSeq
<th colspan="2" align="right"><code>SwizSeq
<tr>
<th colspan="3" align="right"><code>Array
<th colspan="3" align="right"><code>Array
<tr>
<th colspan="4" align="right"><code>ContArray
<td><sub>—/l
<th colspan="4" align="right"><code>ContArray
<tr>
</thead>
<tbody>
<tr>
<th rowspan="4" align="left"><code>flatMap
<td align="right">25 <td><td rowspan="2"><td rowspan="3">
<td><sub><em>4.4
<td align="right">109 <td><td rowspan="2"><td rowspan="3">
<tr>
<td colspan="2" align="right">118
<td><sub><em>2.1
<td colspan="2" align="right">252
<tr>
<td colspan="3" align="right">2092
<td><sub><strong>0.1
<td colspan="3" align="right">208
<tr>
<td colspan="4" align="right">2118
<td><sub><strong>0.1
<td colspan="4" align="right">198
<tr>
<td align="right"><sup>m.j/fM
<td align="right"><sup>1
<td align="right"><sup>1
<td align="right"><sup>1
<td align="right"><sup>1
<td>
<td align="right"><sup>1
<td align="right"><sup><strong>0.7
<td align="right"><sup><em>16
<td align="right"><sup><em>16
<tr>
<th rowspan="4" align="left"><code>map.joined
<td align="right">25 <td><td rowspan="2"><td rowspan="3">
<td><sub><em>4.2
<td align="right">104 <td><td rowspan="2"><td rowspan="3">
<tr>
<td colspan="2" align="right">118
<td><sub><em>1.5
<td colspan="2" align="right">173
<tr>
<td colspan="3" align="right">2094
<td><sub><em>1.5
<td colspan="3" align="right">3240
<tr>
<td colspan="4" align="right">2118
<td><sub><em>1.5
<td colspan="4" align="right">3146
</tbody>
<thead><tr><td><th colspan="10"><code>Unsafe</thead>
<tbody>
<tr><th align="left"><code>BytesReserve<td align="right">70<td colspan="9" rowspan="6">
<tr><th align="left"><code>Bytes<td align="right">77
<tr><th align="left"><code>UInt32InitSeq<td align="right">99
<tr><th align="left"><code>ColorValInitSeq<td align="right">157
<tr><th align="left"><code>FlatMapArray<td align="right">189
<tr><th align="left"><code>FlatMapColorVal<td align="right">198
</tbody>
<thead><tr><td><th colspan="10"><code>—</thead>
<tbody>
<tr><th align="left"><code>for-in.Reserve<td align="right">220<td colspan="9" rowspan="2">
<tr><th align="left"><code>for-in<td align="right">223
</tbody>
</table>

I was motivated to find ever faster implementations, because on my 2008 MBP with 2.4 GHz Intel Core 2 Duo the performance of `flatMap.Array` was slightly worse than the imperative baseline `for-in` (491 vs. 444 μs). It looks like the 10 years of marginal improvements after [hitting performance wall](https://www.technologyreview.com/s/421186/why-cpus-arent-getting-any-faster/), the Intel Xeon E5 2.7 GHz manages to about double the performance to 208 vs. 220 μs for those two cases. Which might mislead us to think that our `flatMap` is doing well… But that's just beefy hardware masking poor code!

From my [previous explorations](https://forums.swift.org/t/troubling-sequence/13130) I knew that using `.lazy` implementations could lead to much faster results. But to my great surprise the `Array`, as well as `ContiguousArray` present some kind of optimization barrier for the swift compiler, because `switching to `.lazy` resulted in **10x degradation** in performance! TODO: Bug Report.

After exploring several `Unsafe` approaches that yielded incremental improvements, I've [stumbled upon promising strategy](https://github.com/apple/swift/pull/20552#discussion_r234764310) of conforming the underlying type to `Sequence` protocol and the absolute champion: `Collection` protocol conformance, which beats even the best Unsafe.BytesReserve variant by a wide margin.

I believe this demonstrates two things:
1) Swift has a **great potential** to leverage Protocol Oriented Programming into being a [Sufficiently Smart Compiler](http://wiki.c2.com/?SufficientlySmartCompiler) and be [as fast as C](http://wiki.c2.com/?AsFastAsCee)*,
2) but its current implementation, even within the narrowest confines of the _most optimized collection_ in the whole Swift Standard Library can accidentally create a **spectacular de-optimization** that show that Swift's current **performance is frustratingly unpredictable**.

*<sub> Not an actual comparison to C. This benchmarks just demonstrates gaping performance differences within Swift Standard Library.

The `ColorRef` benchmarks show the impact of adding one level of indirection and introduction of reference counting.

<table>
<caption><h3 align="center"><code>ColorRef
<thead>
<tr><td rowspan="5"><th colspan="4"><code>lazy<th><th colspan="4"><code>—
<tr>
<th align="right"><code>SwizCol <th><th rowspan="2"><th rowspan="3">
<th rowspan="3">
<th align="right"><code>SwizCol <th><th rowspan="2"><th rowspan="3">
<tr>
<th colspan="2" align="right"><code>SwizSeq
<th colspan="2" align="right"><code>SwizSeq
<tr>
<th colspan="3" align="right"><code>Array
<th colspan="3" align="right"><code>Array
<tr>
<th colspan="4" align="right"><code>ContArray
<td><sub>—/l
<th colspan="4" align="right"><code>ContArray
<tr>
</thead>
<tbody>
<tr>
<th rowspan="4" align="left"><code>flatMap
<td align="right">289 <td><td rowspan="2"><td rowspan="3">
<td><sub><em>1.8
<td align="right">515 <td><td rowspan="2"><td rowspan="3">
<tr>
<td colspan="2" align="right">361
<td><sub><em>1.6
<td colspan="2" align="right">563
<tr>
<td colspan="3" align="right">2399
<td><sub><strong>0.2
<td colspan="3" align="right">428
<tr>
<td colspan="4" align="right">2364
<td><sub><strong>0.2
<td colspan="4" align="right">427
<tr>
<td align="right"><sup>m.j/fM
<td align="right"><sup>1
<td align="right"><sup>1
<td align="right"><sup>1
<td align="right"><sup>1
<td>
<td align="right"><sup>1
<td align="right"><sup>1.2
<td align="right"><sup><em>8
<td align="right"><sup><em>8
<tr>
<th rowspan="4" align="left"><code>map.joined
<td align="right">289 <td><td rowspan="2"><td rowspan="3">
<td><sub><em>2.0
<td align="right">574 <td><td rowspan="2"><td rowspan="3">
<tr>
<td colspan="2" align="right">361
<td><sub><em>1.8
<td colspan="2" align="right">654
<tr>
<td colspan="3" align="right">2390
<td><sub><em>1.8
<td colspan="3" align="right">3493
<tr>
<td colspan="4" align="right">2364
<td><sub><strong>1.4
<td colspan="4" align="right">3419
</tbody>
<thead><tr><td><th colspan="10"><code>—</thead>
<tbody>
<tr><th align="left"><code>for-in.Reserve<td align="right">374 <td colspan="9" rowspan="2">
<tr><th align="left"><code>for-in<td align="right">381
</tbody>
</table>

Here's the relative comparison of value types to reference types in this benchmark family.

<table>
<caption><h3><code>ColorRef</code>/<code>ColorVal
<thead>
<tr><td><th><code>lazy <td><code>— <td>
</thead>
<tbody>
<tr>
<th rowspan="4" align="left"><code>flatMap 
<td align="right"><sup><em>12 
<td align="right"><sup><em>4.7 
<th align="left"><code>SwizCol
<tr>
<td align="right"><sup><em>3.1 
<td align="right"><sup><em>2.2 
<th align="left"><code>SwizSeq
<tr>
<td align="right"><sup>1.1 
<td align="right"><sup><em>2.1 
<th align="left"><code>Array
<tr>
<td align="right"><sup><em>1.1 
<td align="right"><sup><em>2.2 
<th align="left"><code>ContArray
<tr>
<th rowspan="4" align="left"><code>map.joined 
<td align="right"><sup><em>12 
<td align="right"><sup><em>5.5 
<th align="left"><code>SwizCol
<tr>
<td align="right"><sup><em>3.1 
<td align="right"><sup><em>3.8 
<th align="left"><code>SwizSeq
<tr>
<td align="right"><sup>1.1 
<td align="right"><sup>1.1 
<th align="left"><code>Array
<tr>
<td align="right"><sup>1.1 
<td align="right"><sup>1.1
<th align="left"><code>ContArray
</tbody>
<thead><tr><td><th colspan="2"><code>— <td></thead>
<tbody>
<tr><th align="left"><code>for-in.Reserve<td colspan="2" align="right"><sup><em>1,7 <td rowspan="2">
<tr><th align="left"><code>for-in<td colspan="2" align="right"><sup><em>1,7
</tbody>
</table>

palimondo committed:
Merge branch 'master' into fluctuation-of-the-pupil

palimondo commented:
@swift-ci smoke test

palimondo commented:
@swift-ci smoke test

palimondo commented:
@swift-ci smoke test os x

palimondo commented:
@swift-ci smoke test os x

palimondo commented:
@swift-ci smoke test os x

palimondo commented:
@swift-ci smoke test os x

palimondo commented:
@swift-ci smoke test os x

palimondo committed:
Merge branch 'master' into fluctuation-of-the-pupil

palimondo committed:
Merge branch 'master' into fluctuation-of-the-pupil

CodaFi commented:
@palimondo are you still interested in pursuing this change?

CodaFi commented:
It's been a while.  I'm going to close this out due to age and inactivity.  @palimondo The tests here are still valuable to have.  If you find the time, please either reopen this or shoot us another pull request.

palimondo commented:
I’ll get back to this. Sorry for the long hibernation, I have somehow missed the previous ping here.
